contracts/external/cw-payroll/src/contract.rs

```diff
@@ -0,0 +1,599 @@
+    if (VALIDATORS_REWARDS.may_load(deps.storage, &dst_validator)?).is_none() {
+        VALIDATORS_REWARDS.save(deps.storage, &dst_validator, &ValidatorRewards::default())?;
+    }
+
+    // Create message to delegate the underlying tokens
+    let msg = StakingMsg::Redelegate {
+        src_validator,
+        dst_validator,
+        amount: Coin { denom, amount },
+    };
+
+    Ok(Response::new()
+        .add_attribute("method", "redelegate")
+        .add_attribute("amount", amount.to_string())
+        .add_message(msg))
+}
+
+pub fn execute_undelegate(
+    _env: Env,
+    deps: DepsMut,
+    info: MessageInfo,
+    validator: String,
+    amount: Uint128,
+) -> Result<Response, ContractError> {
+    let vesting_payment = VESTING_PAYMENT.load(deps.storage)?;
+
+    // Check sender is the recipient of the vesting payment
+    if info.sender != vesting_payment.recipient {
+        return Err(ContractError::Unauthorized);
+    }
+
+    // Load delegation for specified validator
+    let delegations = STAKED_VESTING_BY_VALIDATOR
+        .may_load(deps.storage, &validator)?
+        .ok_or(ContractError::NoDelegationsForValidator {})?;
+
+    // Load rewards info for specified validator
+    let validator_rewards = VALIDATORS_REWARDS.load(deps.storage, &validator)?;
+
+    // Update vesting payment with amount of staked rewards
+    VESTING_PAYMENT.update(deps.storage, |mut vp| -> Result<_, ContractError> {
+        // Check unstake amount is not greater than vesting payment amount
+        if amount > vp.staked_amount {
+            return Err(ContractError::NotEnoughFunds {});
+        }
+
+        // Update pending rewards
+        vp.calc_pending_rewards(validator_rewards.rewards_per_token, delegations)?;
+
+        // Update amounts
+        vp.staked_amount -= amount;
+        Ok(vp)
+    })?;
+
+    let denom = deps.querier.query_bonded_denom()?;
+
+    // Create message to delegate the underlying tokens
+    let msg = StakingMsg::Undelegate {
+        validator: validator.clone(),
+        amount: Coin { denom, amount },
+    };
+
+    // Update amount delegated to validator
+    STAKED_VESTING_BY_VALIDATOR.save(deps.storage, &validator, &(delegations - amount))?;
+
+    Ok(Response::default().add_message(msg))
+}
+
+pub fn execute_withdraw_rewards(
+    env: Env,
+    deps: DepsMut,
+    validator: String,
+) -> Result<Response, ContractError> {
+    // Query full delegation to get the total rewards amount
+    let delegation_query = deps
+        .querier
+        .query_delegation(env.contract.address, validator.clone())?;
+
+    // Total rewards we have from this validator
+    let total_accumulated_rewards = &match delegation_query {
+        Some(delegation) => delegation.accumulated_rewards,
+        None => return Err(ContractError::NoDelegationsForValidator {}),
+    };
+
+    // Check to make sure there are rewards
+    if total_accumulated_rewards.is_empty() || total_accumulated_rewards[0].amount.is_zero() {
```

Comment:
are we guaranteed that the first element in this list will be the native staking denomination? if not, we may want to check that here.

Reply:
Not needed anymore with the refactor!